### PR TITLE
update makefile dependences

### PR DESCRIPTION
**Pull request by sebpop**

### CODE REVIEW

**Comment by Dead2** [comment]
Hmm, it looks like this removes a whole lot more than it adds back. Is that correct?
And it looks like you didn't use 'make depend' to generate this, if not, how did you generate it?

**Comment by sebpop** [comment]
I tried to use `make depend` and it doesn't work, so I updated the dependencies by hand.
When you run `makedepend` it would only generate the include file dependencies for the specific configuration enable by configure: i.e., on an x86 configure arm specific files would not be generated.

**Comment by sebpop** [comment]
I added code to `configure` to automatically compute the dependencies for each `.c` file.

File: configure [hunk 1]
```diff
@@ -1301,6 +1301,47 @@ sed < $SRCDIR/Makefile.in "
 /^ARCH_SHARED_OBJS *=/s#=.*#=$ARCH_SHARED_OBJS#
 " > Makefile
 
+# Append header files dependences.
+for file in $(ls -1 *.c test/*.c test/fuzz/*.c $ARCHDIR/*.c); do
```
Comment (mtl1979) [inline configure]: Does this work if we are configuring outside source directory?

File: Makefile.in [hunk 2]
```diff
@@ -181,16 +181,16 @@ $(STATICLIB): $(OBJS)
...
-example.o: $(SRCDIR)/test/example.c $(SRCDIR)/zlib$(SUFFIX).h zconf$(SUFFIX).h
+example.o:
 	$(CC) $(CFLAGS) $(INCLUDES) -c -o $@ $(SRCDIR)/test/example.c
 
-minigzip.o: $(SRCDIR)/test/minigzip.c $(SRCDIR)/zlib$(SUFFIX).h zconf$(SUFFIX).h
+minigzip.o:
 	$(CC) $(CFLAGS) $(INCLUDES) -c -o $@ $(SRCDIR)/test/minigzip.c
 
-example64.o: $(SRCDIR)/test/example.c $(SRCDIR)/zlib$(SUFFIX).h zconf$(SUFFIX).h
+example64.o: $(SRCDIR)/test/example.c $(SRCDIR)/zbuild.h $(SRCDIR)/zlib.h $(SRCDIR)/zlib-ng.h
```
Comment (mtl1979) [inline Makefile.in]: This doesn't look right...

File: Makefile.in [hunk 3]
```diff
@@ -181,16 +181,16 @@ $(STATICLIB): $(OBJS)
...
 	$(CC) $(CFLAGS) -D_FILE_OFFSET_BITS=64 $(INCLUDES) -c -o $@ $(SRCDIR)/test/example.c
 
-minigzip64.o: $(SRCDIR)/test/minigzip.c $(SRCDIR)/zlib$(SUFFIX).h zconf$(SUFFIX).h
+minigzip64.o: $(SRCDIR)/test/minigzip.c $(SRCDIR)/zbuild.h $(SRCDIR)/zlib.h $(SRCDIR)/zlib-ng.h
```
Comment (mtl1979) [inline Makefile.in]: Doesn't look right either...

File: Makefile.in (hunk 2 above)
Comment (sebpop) [reply Makefile.in]: I added a special case for this one and the one below.

**Comment by sebpop** [comment]
I tested the updated patch with a build in tree and a build in a separate dir.

**Comment by Dead2** [comment]
Rebase please :)

**Comment by Dead2** [comment]
Uhm, did it fail? Looks like all dependencies were removed now.

**Comment by sebpop** [comment]
I intentionally removed all the dependencies, as they get generated by the configure script from Makefile.in to Makefile.

**Comment by Dead2** [comment]
Question is, does this make problems for those who compile zlib-ng as part of their own project, and do not run configure? I know some run cmake, but probably not all.

**Comment by mtl1979** [comment]
Not running configure or cmake is error by itself and can cause more serious issues than creating dependency information inside configure script instead of hard-coding it in Makefile.in.

When you do repeated builds, the dependency information is used to decide what files need to be rebuilt even if timestamp of that specific file hasn't changed.

Until now, a lot of the dependency information has been incomplete due to constant changing of files that are included by the .c files in zlib-ng.

**Comment by Dead2** [comment]
That does not change the fact that lots of projects compile zlib statically without running configure or cmake properly.

**Comment by mtl1979** [comment]
We can't fix stupid even if we smash our heads against wall... But we can make our own code better for people who actually do things correctly...

**Comment by Dead2** [comment]
Having a script that can generate this is great, but I still think we should have the contents in the files by default. The script can be a maintainer-type script that we run before making releases. It could also be run automatically by configure to make sure the files are up-to-date.

That also makes the changes easier to track in git. Optimally we would also have a travis/buildkite script that checks whether a PR contains the updated deps and fails if it did not.

Or we could require people use cmake subprojects, except I have still not found a good way to have cmake subprojects actually work properly (ex: not only using default values for all options but actually running the scripts to check stuff).

**Comment by mtl1979** [comment]
@Dead2 If we want to go all-in, we should have git pre-commit hook that refreshes Makefile.in dependency information automatically... As Git is MinGW executable on Windows, it will work even when using GUI version.

**Comment by Dead2** [comment]
Oh nice, didn't think of that. Would the git repo inherit that from github on clone though, or does it need to be manually set up?

**Comment by mtl1979** [comment]
I've always manually set it up, but it's possible to set it up inside configure. Basically after you refresh the dependencies in Makefile files, you copy the dependencies to equivalent Makefile.in and replace values of SRCDIR and SRCTOP with the variable names. You will also need to replace ```zlib.h``` and ```zlib-ng.h``` with ```zlib$(SUFFIX).h```, and also ```zconf.h``` and ```zconf-ng.h``` with ```zconf$(SUFFIX).h```.

**Comment by Dead2** [comment]
Needs a rebase

**Pull request merged**